_posts/2021-07-13-ci-metrics.md

```diff
@@ -0,0 +1,116 @@
+
+This approach is called an *error budget* and descibed in all of its details and consequences in the free Google book [Site Reliability Engineering](https://sre.google/sre-book/table-of-contents/). This principle is mostly applied for services towards external customers, such as a web application or cloud engine. But we realized that it equally well applies to our development team as an internal service -- we are just our own customers, with two different hats: Most of the time we develop Cockpit, but sometimes we switch roles and develop our infrastructure.
+
+## Define Service Level Objectives
+
+The first step of that story is to sit down with the stakeholders -- in this case our own development team. We discussed our high-level expectations on our development process, and agreed that we are "happy" as long as these hold:
+
+ - PRs get validated in a reasonable time (queue + test run time).
+ - We don't waste time on interpreting unstable test results.
+ - We are not afraid of touching code.
+ - Test failures are relevant and meaningful. Relieve us from having to decide about “unrelated or not” every. single. time.
+
+These expectations are not yet objective; they need to be translated into concrete and measurable "Service Level Indicators" (SLI), together with a definition which range of an SLI we consider as "good"/"keeps us happy" according to the abstract expectations above. The latter is called a [Service Level Objective (SLO)](https://en.wikipedia.org/wiki/Service-level_objective). For us they look like this, in descending importance:
+
+
+ 1. A merged PR becomes fully green with a 75% chance at the first attempt, and with a 95% chance after one retry.
+ 2. Every individual test succeeds at least 90% of the time.
+ 3. 95% of all PRs are merged without failed tests.
+ 4. 95% of test runs take no more than 1 hour to execute.
+ 5. 95% of test runs spend no more than 5 minutes in the queue until they get assigned to a runner.
+ 6. 95% of scheduled tests run through to completion (all tests ran and status got reported to PR).
+
+ These levels are not very ambitious compared to [five nines](https://en.wikipedia.org/wiki/High_availability#Percentage_calculation) that are often required in other places. But first, we intentionally started low to get a grasp of how an error budget works; and second, when we started this we were rather far away from even meeting these. These SLOs represented a challenge which was achievable with a realistic amount of effort, while being good enough to not hamper our Cockpit development life.
```

Comment:
I would suggest dropping the second ... you already cover the point in the next sentence.

Reply:
Well spotted, dropped.